### PR TITLE
[debug mode] Support dispatching into subgraphs in DebugMode for InvokeSubgraph

### PR DESCRIPTION
**Pull request by yushangdi**
Used to debug invoke_subgraph gradient accuracy in https://github.com/pytorch/pytorch/pull/170485

```
python test/distributed/tensor/debug/test_debug_mode.py -k test_invoke_subgraph
```

We need to 
1) clearly mark the beginning and end of invoke_subgraph
2) actually enable debug_mode for the subgraph call in invoke_subgraph

Example debug string:

```
    def test_nested_invoke_subgraph(self):
        # Test that DebugMode can trace the operations inside
        # invoke_subgraph HOP
        @torch.compiler.nested_compile_region
        def ggn(x):
            a = torch.cos(x)
            return a

        @torch.compiler.nested_compile_region
        def gn(x):
            x = x + 1
            x = ggn(x)
            a = torch.sin(x)
            return a

        def fn(x):
            foo1 = gn(x)
            y = foo1 * 2
            return y

            """\
    torch.ops.higher_order.invoke_subgraph(subgraph_1, t: f32[8, 8])  ->  ('t: f32[8, 8]',)
    [annotate] [enter InvokeSubgraph HOP] subgraph_1
      aten::add.Tensor(t: f32[8, 8], 1)  ->  t: f32[8, 8]
      torch.ops.higher_order.invoke_subgraph(subgraph_0, t: f32[8, 8])  ->  ('t: f32[8, 8]',)
      [annotate] [enter InvokeSubgraph HOP] subgraph_0
        aten::cos(t: f32[8, 8])  ->  t: f32[8, 8]
      [annotate] [exit InvokeSubgraph HOP] subgraph_0
      aten::sin(t: f32[8, 8])  ->  t: f32[8, 8]
    [annotate] [exit InvokeSubgraph HOP] subgraph_1
    aten::mul.Tensor(t: f32[8, 8], 2)  ->  t: f32[8, 8]""",  # noqa: B950

```

### CODE REVIEW

File: torch/_higher_order_ops/invoke_subgraph.py [hunk 1]
```diff
@@ -169,6 +174,39 @@ def gen_schema(self, subgraph, identifier, *operands):
...
+@register_debug_mode_log_hook
+def invoke_subgraph_dispatch_hook(func, types, args, kwargs, result, debug_mode):
+    if func is torch._higher_order_ops.invoke_subgraph:
+        subgraph_name = args[1]
+        debug_mode.call_depth += 1
+        debug_mode._handle_annotate(f"[exit InvokeSubgraph HOP] {subgraph_name}")
+        debug_mode.call_depth -= 1
+        global _invoke_subgraph_debug_mode_stack
+        _invoke_subgraph_debug_mode_stack.pop()
+        return None
```
Comment (pianpwk) [inline torch/_higher_order_ops/invoke_subgraph.py]: hm, is it not possible to put a `get_active_debug_mode()` somewhere in invoke_subgraph, instead of adding hooks? It seems like the flow currently is: for every __torch_dispatch__, check before & after to see if we need to run this hook.

I see your comment about losing DebugMode - is that why we can't call `get_active_debug_mode()` here instead of all the threading?

Comment (yushangdi) [reply torch/_higher_order_ops/invoke_subgraph.py]: Yes. When we call subgraph, it's after we have dispatched in DebugMode, so we don't have DebugMode in the dispatch stack anymore

Comment (yushangdi) [reply torch/_higher_order_ops/invoke_subgraph.py]: Specially, we're inside the `func` call here: https://github.com/pytorch/pytorch/blob/d765bcdbf67ec9fe2d5e017460b98fa8402bb955/torch/utils/_debug_mode.py#L900

the invoke_subgraph HOP is already dispatched in debug mode

File: torch/_higher_order_ops/invoke_subgraph.py [hunk 2]
```diff
@@ -169,6 +174,39 @@ def gen_schema(self, subgraph, identifier, *operands):
...
+        debug_mode.call_depth += 1
+        debug_mode._handle_annotate(f"[enter InvokeSubgraph HOP] {subgraph_name}")
+        debug_mode.call_depth -= 1
+        global _invoke_subgraph_debug_mode_stack
+        _invoke_subgraph_debug_mode_stack.append(debug_mode)
+        return None
+
+
+@register_debug_mode_log_hook
+def invoke_subgraph_dispatch_hook(func, types, args, kwargs, result, debug_mode):
```
Comment (zou3519) [inline torch/_higher_order_ops/invoke_subgraph.py]: I"m confused, DebugMode is a TorchDispatchMode. There is an official mechanism to get a HOP to work with a TorchDispatchMode, you can use redirect_to_mode, or use a .py_impl(DebugMode). Why don't those work?

Comment (yushangdi) [reply torch/_higher_order_ops/invoke_subgraph.py]: `redirect_to_mode` only dispatch the `invoke_subgraph` HOP node itself to the DebugMode, but not the calls for subgraphs in the HOP. Have not tried `.py_impl(DebugMode)`, trying...

Comment (zou3519) [reply torch/_higher_order_ops/invoke_subgraph.py]: I think I don't understand what exactly you're trying to do here

Comment (bdhirsh) [reply torch/_higher_order_ops/invoke_subgraph.py]: @yushangdi specifically, I think Richard is pointing out that:

(1) you are adding some extra logic to handle invoke_subgraph by manually registering these extra "hooks", which you call at the beginning and end of your `DebugMode.__torch_dispatch__`, and manually tweaking the "dense" impl of `invoke_subgraph` to push `debug_mode` back onto the mode stack if there were any previously

(2) Ideally, though, InvokeSubgraph shouldn't need to know anything about DebugMode. And there are existing APIs that allow you to ensure this. Specifically, you can write an `invoke_subgraph.py_impl(DebugMode)` implementation (that can live in `_debug_mode.py`, so invoke_subgraph doesn't need to know anything about it), which just does something like:
```
def invoke_subgraph_debug_mode_impl(debug_mode, *args, **kwargs):
    # ... some set up to tweak indenting for the ops inside of the subgraph
    with debug_mode:
        return invoke_subgraph(*args, **kwargs)
```

Comment (yushangdi) [reply torch/_higher_order_ops/invoke_subgraph.py]: yeah I think invoke_subgraph.py_impl(DebugMode) should work. by "trying" I just meant I'm trying to re-implement with `py_impl`

File: torch/_higher_order_ops/invoke_subgraph.py [hunk 3]
```diff
@@ -675,12 +676,64 @@ def autograd_fn_callable(*args):
...
+@invoke_subgraph.py_impl(DebugMode)
+def _(debug_mode, subgraph, identifier, *operands):
+    from torch.utils._python_dispatch import _get_current_dispatch_mode
+
+    mode = _get_current_dispatch_mode()
+    assert mode is None, "Mode should never be enabled for CPU/CUDA key"
+
+    # If the HOP is dispatched from DebugMode, we should enable debug_mode
+    # for the subgraph call.
+    ctx = _push_mode_temporarily(debug_mode)
```
Comment (bdhirsh) [inline torch/_higher_order_ops/invoke_subgraph.py]: for my understanding - why can't you just use `with debug_mode:` instead of this API? If there is some extra stuff in `DebugMode.__enter__` that you are explicitly trying to skip, maybe we're doing something wrong?

Comment (zou3519) [reply torch/_higher_order_ops/invoke_subgraph.py]: +1

**Review by zou3519 — COMMENTED** [review]
code looks a lot better than before, just wondering about Brian's comment

File: torch/_higher_order_ops/invoke_subgraph.py (hunk 3 above)
Comment (yushangdi) [reply torch/_higher_order_ops/invoke_subgraph.py]: that's because in DebugMode's __enter__, there is a `reset` that clears all existing entries in DebugMode:
https://github.com/pytorch/pytorch/blob/056d2635c696b536c8a3d4092a7983f53100291a/torch/utils/_debug_mode.py#L748

If we remove the `reset`, we can just do `with debugmode`. 

@pianpwk do you know why we had the `reset` in __enter__ and can we remove it?

Comment (pianpwk) [reply torch/_higher_order_ops/invoke_subgraph.py]: That's my mistake, but we should probably split reset into 1) clearing logs and tensor tracking (we shouldn't do that on enter) 2) resetting call depth & other fx metadata (we probably should do on enter/exit)

Comment (yushangdi) [reply torch/_higher_order_ops/invoke_subgraph.py]: removed now!

**Comment by yushangdi** [comment]
> code looks a lot better than before, just wondering about Brian's comment

@zou3519 @bdhirsh  thanks for the review! changed to `with debugmode` now after removing `reset()` in DebugMode's enter context.

File: torch/_higher_order_ops/invoke_subgraph.py [hunk 4]
```diff
@@ -675,12 +676,47 @@ def autograd_fn_callable(*args):
     return autograd_fn_callable(*operands)
 
 
+@invoke_subgraph.py_impl(DebugMode)
+def _(debug_mode, subgraph, identifier, *operands):
+    from torch.utils._python_dispatch import _get_current_dispatch_mode
+
+    mode = _get_current_dispatch_mode()
+    assert mode is None, "Mode should never be enabled for CPU/CUDA key"
```
Comment (zou3519) [inline torch/_higher_order_ops/invoke_subgraph.py]: what is this assertion lol. this is not the cpu/cuda key

**Review by zou3519 — APPROVED** [review]
the assertion message is wrong, I don't know what you wanted to assert. Other than that this lgtm

File: torch/_higher_order_ops/invoke_subgraph.py (hunk 4 above)
Comment (yushangdi) [reply torch/_higher_order_ops/invoke_subgraph.py]: oh, copied from @invoke_subgraph.py_impl(DispatchKey.CompositeExplicitAutograd), removed now.
